packages/components/src/date-time/time/index.tsx

```diff
@@ -0,0 +1,309 @@
+ * import { useState } from '@wordpress/element';
+ *
+ * const MyTimePicker = () => {
+ *   const [ time, setTime ] = useState( new Date() );
+ *
+ *   return (
+ *     <TimePicker
+ *       currentTime={ date }
+ *       onChange={ ( newTime ) => setTime( newTime ) }
+ *       is12Hour
+ *     />
+ *   );
+ * };
+ * ```
+ */
+export function TimePicker( {
+	is12Hour,
+	currentTime,
+	onChange,
+}: TimePickerProps ) {
+	const [ date, setDate ] = useState( () =>
+		// Truncate the date at the minutes, see: #15495.
+		currentTime ? moment( currentTime ).startOf( 'minutes' ) : moment()
+	);
+
+	// Reset the state when currentTime changed.
+	useEffect( () => {
+		setDate(
+			currentTime ? moment( currentTime ).startOf( 'minutes' ) : moment()
+		);
+	}, [ currentTime ] );
+
+	const { day, month, year, minutes, hours, am } = useMemo(
+		() => ( {
+			day: date.format( 'DD' ),
+			month: date.format( 'MM' ),
+			year: date.format( 'YYYY' ),
+			minutes: date.format( 'mm' ),
+			hours: date.format( is12Hour ? 'hh' : 'HH' ),
+			am: Number( date.format( 'H' ) ) <= 11 ? 'AM' : 'PM',
+		} ),
+		[ date, is12Hour ]
+	);
+
+	/**
+	 * Function that sets the date state and calls the onChange with a new date.
+	 * The date is truncated at the minutes.
+	 *
+	 * @param {Moment} newDate The date object.
+	 */
+	function changeDate( newDate: Moment ) {
+		setDate( newDate );
+		onChange?.( newDate.format( TIMEZONELESS_FORMAT ) );
+	}
+
+	const buildNumberControlChangeCallback = (
+		method: 'hours' | 'minutes' | 'date' | 'year'
+	) => {
+		const callback: InputChangeCallback = ( value, { event } ) => {
+			if ( ! ( event.target instanceof HTMLInputElement ) ) {
+				return;
+			}
+
+			if ( ! event.target.validity.valid ) {
+				return;
+			}
+
+			// We can safely assume value is a number if target is valid.
+			let numberValue = Number( value );
+
+			// If the 12-hour format is being used and the 'PM' period is
+			// selected, then the incoming value (which ranges 1-12) should be
+			// increased by 12 to match the expected 24-hour format.
+			if ( method === 'hours' && is12Hour ) {
+				numberValue = from12hTo24h( numberValue, am === 'PM' );
+			}
+
+			const newDate = date.clone()[ method ]( numberValue );
+			setDate( newDate );
+			onChange?.( newDate.format( TIMEZONELESS_FORMAT ) );
+		};
+		return callback;
+	};
+
+	function buildAmPmChangeCallback( value: 'AM' | 'PM' ) {
+		return () => {
+			if ( am === value ) {
+				return;
+			}
+
+			const parsedHours = parseInt( hours, 10 );
+
+			const newDate = date
+				.clone()
+				.hours( from12hTo24h( parsedHours, value === 'PM' ) );
+
+			changeDate( newDate );
+		};
+	}
+
+	const dayField = (
+		<DayInput
+			className="components-datetime__time-field-day-input"
+			label={ __( 'Day' ) }
+			hideLabelFromVision
+			__next36pxDefaultSize
+			value={ day }
+			step={ 1 }
+			min={ 1 }
+			max={ 31 }
+			hideHTMLArrows
+			isPressEnterToChange
+			isDragEnabled={ false }
```

Comment:
Just wondering why we're disabling the drag feature? We don't currently have this behaviour on `trunk`, but I wondered if it'd make it quicker for Day and Time to be able to click and drag it up and down?

Reply:
It's not a great UX, unfortunately, because editing the post date (using the `editPost()` action) is fairly slow and re-rendering the `react-dates` calendar component is _very_ slow. The drag gesture ends up feeling really sluggish. We could spend some time optimising all of that but I don't really think it's worth the investment for now.

Reply:
Good call... perhaps an exploration for an enhancement at some point after this lands!